### ISSUE TITLE
Where is the pain?

### ISSUE THREAD

**Issue by mguterl** [moran742/GeekOnTheRun#1]
Hey @moran742,

The first time I perform a code review, I typically start with a question. Where are you feeling pain in this code? If you're not feeling pain yet because you haven't had to change much, where do you think you will feel pain in the future?

There's no right answer to the questions above and I already have some feedback based on what you have provided, but I want to see your thoughts first.

**Comment by moran742** [comment]
I think I will feel pain when I begin to change the display of the information. I feel like if I begin changing how the data is displayed it will either be a convoluted, time-consuming process or things will break.  I also plan to add more features and am afraid that I'm abusing the routes file with all of the code.  Since I'm a newb, I'm unsure where to put some of the larger blocks of code when a route is called.

**Comment by mguterl** [comment]
Okay, let's start with the large amount of logic in each route. Your instinct here is correct, you shouldn't have too much going on inside of these blocks. You should be delegating down to application specific objects. Another thing to watch out for is you should limit the number of instance variables that your view depends on or else it will make things harder to change. It is also worth noting that the ruby convention is to use underscore_case instead of camelCase.

Let's start with https://github.com/moran742/GeekOnTheRun/blob/master/app.rb#L16

I would recommend only binding a single variable to your view from this action. In this case `@game` or `@game_details`

In the end the route should look something like this:

``` ruby
get '/game/*' do
  @title = "Mobile BGG"
  @game = BoardGameGeek::Game.find params[:splat].first
  haml :game
end
```

I would recommend grabbing this code snippet and drop it into your project: https://raw.github.com/gist/271214/19c5d1f25c0793e8f3fa65573fb595f002622c4c/super_struct.rb

``` ruby
require 'super_struct'

module BoardGameGeek
  class Game < SuperStruct.new(:game_id, :image_id) # need to add the rest of your instance variables here
    def self.find(game_id)
      url = "http://boardgamegeek.com/xmlapi2/thing?type=boardgame&stats=1&id=" + game_id
      document = Nokogiri::XML(open(url))

      attributes = {
        :game_id => doc.xpath("//item/@id").text,
        :thumbnail => doc.xpath("//thumbnail").text.match(/\d+/).to_s,
        # other attributes from game route
      }

      Game.new(attributes)
    end
  end
end
```

Extracting this logic into it's own class helps make everything easy to test and encapsulates the idea of a game into a single object instead of a large number of variables bound to your view. Let me know how this first refactoring goes and I'll make more suggestions once this part is cleaned up.

**Comment by mguterl** [comment]
I should mention, you'll also need to change your view to reference `@game` and it's methods (attributes) rather than the instance variables.

**Comment by moran742** [comment]
Wow. You are the man.  This is really good stuff!  I will work on it and give you a shout.  I really appreciate your taking the time to school me.  Thanks.

**Comment by moran742** [comment]
I skipped the drink up tonight to try and understand this!  I did not clearly understand what a Struct was prior to this, but now I think I understand:
- It's a way to bundle attributes of an object together?  
- And I think I understand that the super_struct allows us to very easily slide in a hash without having to worry about what position each value is in?  
- I don't think I'm following the self.find method name though.  Why not just name it "initialize"?

Man, I would disagree with anyone that thinks Ruby is a "simple" language.

**Comment by mguterl** [comment]
Yep, a struct is just a simple wrapper for a bucket of attributes. Compare the following examples:

``` ruby
class Person < Struct.new(:name, :age)
end

class Person

  attr_accessor :name, :age  

  def initialize(name, age)
    @name, @age = name, age
  end

end
```

Both classes above are equivalent.

You are also correct about the SuperStruct. It is basically the same as a Struct, but it makes the initializer much simpler, replacing connascence of position with connascence of name.

``` ruby
class Person < Struct.new(:name, :age)
  def initialize(attributes)
    attributes.each do |attr, value|
      send("#{attr}=", value) if respond_to?("#{attr}=")
    end
  end
end
```

The above example is roughly equivalent to the SuperStruct.

The BoardGameGeek::Game object might want to be initialized without parsing it from the XML API. This is one of the reasons for creating a separate find method instead of just using initialize. In practice either will work, but I find it best to separate the concern of finding, parsing, and initializing an object from just initializing it.

Does that help?

**Comment by mguterl** [comment]
Hey @moran742 if you push what you have to a branch, I can help debug why your require is not working.

**Comment by moran742** [comment]
Thanks for keeping me on task.  I've got the require working after we talked on Friday...duh on my part.  Working on the view.  Hope to push the changes shortly.

**Comment by moran742** [comment]
Sweet.  That went well.  My next ToDo's:
- Refactor the rest of the code in the same way as was done for the Game class and view.  
- Write tests

**Comment by mguterl** [comment]
Awesome, I would recommend flipping the order of those tasks and starting with some high level tests first, then refactor. That will help give you some confidence when you refactor. I know that sometimes starting to write tests in a project is the hardest part so I'll throw some code your way later today showing you my idea in more concrete terms.
